### ISSUE TITLE
Parallel running nodes become parent/children dependency in Argo 1.3.0

### ISSUE THREAD

**Issue by zijianjoy** [argoproj/argo-workflows#6267]
## Summary 

After upgrading from Argo 2.19.0 to 3.1.0, two parallel running nodes have become parent/children dependency as shown in `status->nodes` path. Using Argo official example for a diamond workflow: https://argoproj.github.io/argo-workflows/examples/#dag.

The dependency of dag should look like:
```
         A
       /   \
     B       C
       \   /
         D
```

But as a result of running against Argo 3.1.0, it is actually:

```
         A
       /   
     B   -  C 
           /
         D
```

Both nodes (B and C) are running in parallel and don't have dependency on each other, but the value from yaml file doesn't show the correct relationship.

Reference for a real issue: https://github.com/kubeflow/pipelines/issues/5944.

## Diagnostics

Reproduce steps:

1. Copy the content from https://argoproj.github.io/argo-workflows/examples/#dag and save in a `diamond.yaml` file. Set the name to be 
2. `kubectl apply -f diamond.yaml ` to apply workflow with name `dag-diamond-2`.
3. Once workflow finishes, call `kubectl get workflow.argoproj.io/dag-diamond-2  -oyaml > dag-status.yaml`
4. This is the content of `dag-status.yaml`:
Checkout `status: nodes: {node-name}: children`
Note that Node A (dag-diamond-2-2584207632) only has one child: B (dag-diamond-2-2634540489), and Node B has a one child: C (dag-diamond-2-2617762870). But A should have two children B and C. Also B and C should have one common child which is D.
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"argoproj.io/v1alpha1","kind":"Workflow","metadata":{"annotations":{},"name":"dag-diamond-2","namespace":"kubeflow"},"spec":{"entrypoint":"diamond","serviceAccountName":"pipeline-runner","templates":[{"container":{"command":["echo","{{inputs.parameters.message}}"],"image":"alpine:3.7"},"inputs":{"parameters":[{"name":"message"}]},"name":"echo"},{"dag":{"tasks":[{"arguments":{"parameters":[{"name":"message","value":"A"}]},"name":"A","template":"echo"},{"arguments":{"parameters":[{"name":"message","value":"B"}]},"dependencies":["A"],"name":"B","template":"echo"},{"arguments":{"parameters":[{"name":"message","value":"C"}]},"dependencies":["A"],"name":"C","template":"echo"},{"arguments":{"parameters":[{"name":"message","value":"D"}]},"dependencies":["B","C"],"name":"D","template":"echo"}]},"name":"diamond"}]}}
  creationTimestamp: "2021-07-01T21:18:41Z"
  generation: 5
  labels:
    workflows.argoproj.io/completed: "true"
    workflows.argoproj.io/phase: Succeeded
  name: dag-diamond-2
  namespace: kubeflow
  resourceVersion: "21880772"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/kubeflow/workflows/dag-diamond-2
  uid: f357ab08-59f2-4409-8ea4-5cbd24f77c7e
spec:
  arguments: {}
  entrypoint: diamond
  serviceAccountName: pipeline-runner
  templates:
  - container:
      command:
      - echo
      - '{{inputs.parameters.message}}'
      image: alpine:3.7
      name: ""
      resources: {}
    inputs:
      parameters:
      - name: message
    metadata: {}
    name: echo
    outputs: {}
  - dag:
      tasks:
      - arguments:
          parameters:
          - name: message
            value: A
        name: A
        template: echo
      - arguments:
          parameters:
          - name: message
            value: B
        dependencies:
        - A
        name: B
        template: echo
      - arguments:
          parameters:
          - name: message
            value: C
        dependencies:
        - A
        name: C
        template: echo
      - arguments:
          parameters:
          - name: message
            value: D
        dependencies:
        - B
        - C
        name: D
        template: echo
    inputs: {}
    metadata: {}
    name: diamond
    outputs: {}
status:
  artifactRepositoryRef:
    default: true
  conditions:
  - status: "False"
    type: PodRunning
  - status: "True"
    type: Completed
  finishedAt: "2021-07-01T21:19:11Z"
  nodes:
    dag-diamond-2:
      children:
      - dag-diamond-2-2584207632
      displayName: dag-diamond-2
      finishedAt: "2021-07-01T21:19:11Z"
      id: dag-diamond-2
      name: dag-diamond-2
      outboundNodes:
      - dag-diamond-2-2668095727
      phase: Succeeded
      progress: 4/4
      resourcesDuration:
        cpu: 4
        memory: 0
      startedAt: "2021-07-01T21:18:41Z"
      templateName: diamond
      templateScope: local/dag-diamond-2
      type: DAG
    dag-diamond-2-2584207632:
      boundaryID: dag-diamond-2
      children:
      - dag-diamond-2-2634540489
      displayName: A
      finishedAt: "2021-07-01T21:18:43Z"
      id: dag-diamond-2-2584207632
      inputs:
        parameters:
        - name: message
          value: A
      name: dag-diamond-2.A
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: artifacts/dag-diamond-2/2021/07/01/dag-diamond-2-2584207632/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 1
        memory: 0
      startedAt: "2021-07-01T21:18:41Z"
      templateName: echo
      templateScope: local/dag-diamond-2
      type: Pod
    dag-diamond-2-2617762870:
      boundaryID: dag-diamond-2
      children:
      - dag-diamond-2-2668095727
      displayName: C
      finishedAt: "2021-07-01T21:18:53Z"
      id: dag-diamond-2-2617762870
      inputs:
        parameters:
        - name: message
          value: C
      name: dag-diamond-2.C
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: artifacts/dag-diamond-2/2021/07/01/dag-diamond-2-2617762870/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 1
        memory: 0
      startedAt: "2021-07-01T21:18:51Z"
      templateName: echo
      templateScope: local/dag-diamond-2
      type: Pod
    dag-diamond-2-2634540489:
      boundaryID: dag-diamond-2
      children:
      - dag-diamond-2-2617762870
      displayName: B
      finishedAt: "2021-07-01T21:18:53Z"
      id: dag-diamond-2-2634540489
      inputs:
        parameters:
        - name: message
          value: B
      name: dag-diamond-2.B
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: artifacts/dag-diamond-2/2021/07/01/dag-diamond-2-2634540489/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 1
        memory: 0
      startedAt: "2021-07-01T21:18:51Z"
      templateName: echo
      templateScope: local/dag-diamond-2
      type: Pod
    dag-diamond-2-2668095727:
      boundaryID: dag-diamond-2
      displayName: D
      finishedAt: "2021-07-01T21:19:03Z"
      id: dag-diamond-2-2668095727
      inputs:
        parameters:
        - name: message
          value: D
      name: dag-diamond-2.D
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: artifacts/dag-diamond-2/2021/07/01/dag-diamond-2-2668095727/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 1
        memory: 0
      startedAt: "2021-07-01T21:19:01Z"
      templateName: echo
      templateScope: local/dag-diamond-2
      type: Pod
  phase: Succeeded
  progress: 4/4
  resourcesDuration:
    cpu: 4
    memory: 0
  startedAt: "2021-07-01T21:18:41Z"
```

👀 Yes! We need all of your diagnostics, please make sure you add it all, otherwise we'll go around in circles asking you for it:

What Kubernetes provider are you using? GKE

What version of Argo Workflows are you running? 3.1.0

What executor are you running? Docker

Did this work in a previous version? I.e. is it a regression? Yes.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by sarabala1979** [comment]
is it happening alway? I tried in latest code. I am not able to reproduce it.
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  creationTimestamp: "2021-07-01T22:25:24Z"
  generateName: dag-diamond-
  generation: 5
  labels:
    workflows.argoproj.io/completed: "true"
    workflows.argoproj.io/phase: Succeeded
  name: dag-diamond-622nr
  namespace: argo
  resourceVersion: "223434"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/argo/workflows/dag-diamond-622nr
  uid: 78fb35fc-d0da-4867-946a-11d8aeb93536
spec:
  arguments: {}
  entrypoint: diamond
  templates:
  - dag:
      tasks:
      - arguments:
          parameters:
          - name: message
            value: A
        name: A
        template: echo
      - arguments:
          parameters:
          - name: message
            value: B
        depends: A
        name: B
        template: echo
      - arguments:
          parameters:
          - name: message
            value: C
        depends: A
        name: C
        template: echo
      - arguments:
          parameters:
          - name: message
            value: D
        depends: B && C
        name: D
        template: echo
    inputs: {}
    metadata: {}
    name: diamond
    outputs: {}
  - container:
      command:
      - echo
      - '{{inputs.parameters.message}}'
      image: alpine:3.7
      name: ""
      resources: {}
    inputs:
      parameters:
      - name: message
    metadata: {}
    name: echo
    outputs: {}
status:
  artifactRepositoryRef:
    configMap: artifact-repositories
    key: default-v1
    namespace: argo
  conditions:
  - status: "False"
    type: PodRunning
  - status: "True"
    type: Completed
  finishedAt: "2021-07-01T22:25:54Z"
  nodes:
    dag-diamond-622nr:
      children:
      - dag-diamond-622nr-2850372080
      displayName: dag-diamond-622nr
      finishedAt: "2021-07-01T22:25:54Z"
      id: dag-diamond-622nr
      name: dag-diamond-622nr
      outboundNodes:
      - dag-diamond-622nr-2934260175
      phase: Succeeded
      progress: 4/4
      resourcesDuration:
        cpu: 20
        memory: 12
      startedAt: "2021-07-01T22:25:24Z"
      templateName: diamond
      templateScope: local/dag-diamond-622nr
      type: DAG
    dag-diamond-622nr-2850372080:
      boundaryID: dag-diamond-622nr
      children:
      - dag-diamond-622nr-2900704937
      - dag-diamond-622nr-2883927318
      displayName: A
      finishedAt: "2021-07-01T22:25:30Z"
      id: dag-diamond-622nr-2850372080
      inputs:
        parameters:
        - name: message
          value: A
      name: dag-diamond-622nr.A
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: dag-diamond-622nr/dag-diamond-622nr-2850372080/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 5
        memory: 3
      startedAt: "2021-07-01T22:25:24Z"
      templateName: echo
      templateScope: local/dag-diamond-622nr
      type: Pod
    dag-diamond-622nr-2883927318:
      boundaryID: dag-diamond-622nr
      children:
      - dag-diamond-622nr-2934260175
      displayName: C
      finishedAt: "2021-07-01T22:25:40Z"
      id: dag-diamond-622nr-2883927318
      inputs:
        parameters:
        - name: message
          value: C
      name: dag-diamond-622nr.C
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: dag-diamond-622nr/dag-diamond-622nr-2883927318/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 5
        memory: 3
      startedAt: "2021-07-01T22:25:34Z"
      templateName: echo
      templateScope: local/dag-diamond-622nr
      type: Pod
    dag-diamond-622nr-2900704937:
      boundaryID: dag-diamond-622nr
      children:
      - dag-diamond-622nr-2934260175
      displayName: B
      finishedAt: "2021-07-01T22:25:40Z"
      id: dag-diamond-622nr-2900704937
      inputs:
        parameters:
        - name: message
          value: B
      name: dag-diamond-622nr.B
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: dag-diamond-622nr/dag-diamond-622nr-2900704937/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 5
        memory: 3
      startedAt: "2021-07-01T22:25:34Z"
      templateName: echo
      templateScope: local/dag-diamond-622nr
      type: Pod
    dag-diamond-622nr-2934260175:
      boundaryID: dag-diamond-622nr
      displayName: D
      finishedAt: "2021-07-01T22:25:50Z"
      id: dag-diamond-622nr-2934260175
      inputs:
        parameters:
        - name: message
          value: D
      name: dag-diamond-622nr.D
      outputs:
        artifacts:
        - name: main-logs
          s3:
            key: dag-diamond-622nr/dag-diamond-622nr-2934260175/main.log
        exitCode: "0"
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 5
        memory: 3
      startedAt: "2021-07-01T22:25:44Z"
      templateName: echo
      templateScope: local/dag-diamond-622nr
      type: Pod
  phase: Succeeded
  progress: 4/4
  resourcesDuration:
    cpu: 20
    memory: 12
  startedAt: "2021-07-01T22:25:24Z"
```

![image](https://user-images.githubusercontent.com/33908564/124196930-4ac59b00-da82-11eb-803e-b5874a493b1f.png)

**Comment by Bobgy** [comment]
Yes, @sarabala1979 it happens stability.

@zijianjoy I just realized we can also verify using argo prebuilt installation in release note, rather than our own installation, maybe we did sth wrong in our manifest.

**Comment by sarabala1979** [comment]
@Bobgy  If it is an installation problem,  Please update the details and close it

**Comment by Bobgy** [comment]
@sarabala1979 I verified using quickstart-mysql for argo v3.1.0 and v3.1.1.

Confirmed this issue happens in v3.1.0 and has been fixed in v3.1.1. I think that's why you weren't able to reproduce.

Can you help closing the issue? I don't have permission for this
